src/legacy/server/logging/rotate/log_rotator.js

```diff
@@ -0,0 +1,316 @@
+
+    // init log file size monitor
+    await this._startLogFileSizeMonitor();
+  }
+
+  stop() {
+    if (!this.running) {
+      return;
+    }
+
+    // cleanup exit listener
+    this._deleteExitListener();
+
+    // stop log file size monitor
+    this._stopLogFileSizeMonitor();
+
+    this.running = false;
+  }
+
+  async _shouldUsePolling() {
+    try {
+      // Setup a test file in order to try the fs env
+      // and understand if we need to usePolling or not
+      const tempFileDir = tmpdir();
+      const tempFile = join(tempFileDir, 'kbn_log_rotation_use_polling_test_file.log');
+
+      await mkdirAsync(tempFileDir, { recursive: true });
+      await writeFileAsync(tempFile, '');
+
+      const testWatcher = fs.watch(tempFile, { persistent: false });
+
+      return new Promise(async (resolve) => {
+        let fallbackTimeout = null;
+
+        const onResolve = once(async (resolve, completeStatus) => {
+          clearTimeout(fallbackTimeout);
+
+          testWatcher.close();
+          await unlinkAsync(tempFile);
+          resolve(completeStatus);
+        });
+
+        fallbackTimeout = setTimeout(async () => {
+          await onResolve(resolve, true);
+        }, 15000);
+
+        testWatcher.on('change', async () => {
+          await onResolve(resolve, false);
+        });
+
+        testWatcher.on('error', async () => {
+          await onResolve(resolve, true);
+        });
+
+        await writeFileAsync(tempFile, 'test');
+      });
+    } catch {
+      return true;
+    }
+  }
+
+  _buildWatchCfg(usePolling = false) {
+    return {
+      ignoreInitial: true,
+      awaitWriteFinish: false,
+      useFsEvents: false,
+      usePolling,
+      interval: this.pollingInterval,
+      binaryInterval: this.pollingInterval,
+      alwaysStat: true,
+      atomic: false
+    };
+  }
+
+  async _startLogFileSizeMonitor() {
+    this.usePolling = await this._shouldUsePolling();
+
+    if (this.usePolling && this.usePolling !== this.config.get('logging.rotate.usePolling')) {
+      this.log(
+        ['warning', 'logging:rotate'],
+        'The current environment does not support `fs.watch`. Falling back to polling using `fs.watchFile`'
+      );
+    }
+
+    this.stalker = chokidar.watch(this.logFilePath, this._buildWatchCfg(this.usePolling));
+    this.stalker.on('change', async (filename, stats) => await this._logFileSizeMonitorHandler.bind(this)(filename, stats));
+  }
+
+  async _logFileSizeMonitorHandler(filename, stats) {
+    if (!filename || !stats) {
+      return;
+    }
+
+    this.logFileSize = stats.size || 0;
+    await this.throttledRotate();
+  }
+
+  _stopLogFileSizeMonitor() {
+    if (!this.stalker) {
+      return;
+    }
+
+    this.stalker.close();
+    clearTimeout(this.stalkerUsePollingPolicyTestTimeout);
+  }
+
+  _createExitListener() {
+    process.on('exit', this.stop.bind(this));
+  }
+
+  _deleteExitListener() {
+    process.removeListener('exit', this.stop);
```

Comment:
`bind` creates a new function on every call. it doesn't remove listener here. you can use `stop = () => ` to have bound method

Reply:
👍